qkeras/quantizers.py

```diff
@@ -439,9 +440,537 @@ def trainable_variables(self):
+
+
+    err_msg = (f"Bit count {self.bits} must exceed "
+              f" {self.integer + self.keep_negative}")
+    if self.bits < self.integer + self.keep_negative:
+      raise ValueError(err_msg)
+
+    # Set default quantization scale
+    self.quantization_scale = self.data_type_scale
+
+    # Set scales for tensor alpha
+    if self.alpha is not None and not self.auto_alpha:
+        self.quantization_scale = self.alpha * self.data_type_scale
+
+  def __call__(self, x):
+    """Core quantization function"""
+
+    # Build if not already built
+    self._build()
+
+    # Data type conversion
+    x = K.cast_to_floatx(x)
+    
+    if self.auto_alpha:
+      # get data-dependent quantization scale
+      quantization_scale = self._get_quantization_scale(x)
+    else:
+      # quantization scale determined by quantizer params, not data
+      # see _set_default_quantization_scale for more info
+      quantization_scale = self.quantization_scale
+
+    scaled_xq = self._scale_clip_and_round(x, quantization_scale)
+    xq = scaled_xq * quantization_scale
+
+    res = x + self.qnoise_factor * (xq - x)
+
+    return res
+  
+  def _scale_clip_and_round(self, x, quantization_scale):
+    """Scale, clip, and round x to an integer value in a limited range
+    Note that the internal shift is needed for 1-bit quantization to ensure 
+    that a sign function is used."""
+
+    # special shifting needed to compute a sign function.
+    shift = self.use_sign_function * 0.5
+
+    clip_min, clip_max = self.clip_bounds
+
+    scaled_x = x / quantization_scale
+    clipped_scaled_x = K.clip(scaled_x, clip_min, clip_max)
+    # Round through to nearest integer, using straight-through estimator 
+    # for gradient computations. 
+    scaled_xq = _round_through(
+      clipped_scaled_x - shift,
+      use_stochastic_rounding=self.use_stochastic_rounding,
+      precision=1.0, # using 1.0 precision so that we round to a nearby integer
+    )
+
+    return scaled_xq + shift
+    
+  def _get_quantization_scale(self, x):
+    """Get quantization_scale, either from self or from input x"""
+
+    # Get the minimum floating point scale that does not clip the max of x
+    # This is the quantization scale for alpha="auto"
+    quantization_scale = self._get_quantization_scale_from_max_data(x)
+
+    if self.alpha == "auto_po2":
+      quantization_scale = self._po2_autoscale(x, quantization_scale)
+
+    # update quantization_scale variable
+    # stop_gradient on quantization_scale to ignore dependence on x
+    self.quantization_scale = tf.stop_gradient(quantization_scale)
+
+    # very important that return value is a tf.Variable with shape None
+    return self.quantization_scale 
+
+  def _get_quantization_scale_from_max_data(self, x):
+    """Get the minimum floating point scale that does not clip the max 
+    of x"""
+
+    axis = _get_scaling_axis(self.scale_axis, tf.rank(x))
+
+    clip_min, clip_max = self.clip_bounds
```

Comment:
It seems that we calculate clip bounds multiple times for each call of the quantizer. Specifically, line 802 and line 839 are both called if auto alpha is used (if I understand correctly)

Should we try to store the result somewhere to reduce computation? Or perhaps it doesn't make much of a difference?

Reply:
Yes, that's correct. The clip bound computation is a scalar computation, so it does not add much to the computation time if x is large. 

We could store the clip bounds, but this would require updating them every time a parameter that they depend on is updated.

At one point, I did this for all "input-independent attributes", which involved a lot of condition parameter updates in setter methods. However, this made the code much less legible and provided negligible performance improvements.

Reply:
Sounds good